### PR TITLE
Add Web Indexer for apibay.org 

### PR DESCRIPTION
**Pull request by masterwok**
This pull request adds a new web indexer for apibay.org (the backing API of The Pirate Bay). This is my first contribution to this repository so please review carefully.

I didn't remove the existing pirate bay indexer but I can do so as part of this PR if requested.

I referenced the existing `thepiratebay.yml` and `RarBG` indexer during development.

### CODE REVIEW

File: src/Jackett.Common/Indexers/ApiBay.cs
Comment (ilike2burnthing) [inline src/Jackett.Common/Indexers/ApiBay.cs:33]:
```suggestion
        private static readonly string MagnetUri =
            $"magnet:?xt=urn:btih:{KeyInfoHash}&tr=udp%3A%2F%2Ftracker.coppersurfer.tk" +
            "%3A6969%2Fannounce&tr=udp%3A%2F%2F9.rarbg.to%3A2920%2Fannounce&tr=udp%3" +
            "A%2F%2Ftracker.opentrackr.org%3A1337&tr=udp%3A%2F%2Ftracker.internetwar" +
            "riors.net%3A1337%2Fannounce&tr=udp%3A%2F%2Ftracker.leechers-paradise.or" +
            "g%3A6969%2Fannounce&tr=udp%3A%2F%2Ftracker.coppersurfer.tk%3A6969%2Fann" +
            "ounce&tr=udp%3A%2F%2Ftracker.pirateparty.gr%3A6969%2Fannounce&tr=udp%3A" +
            "%2F%2Ftracker.cyberia.is%3A6969%2Fannounce";
```
Otherwise it names all magnet downloads `Carey M.​ Tribe of Hackers.​ Cybersecurity Advice.​.​.​2019`

**Comment by ilike2burnthing** [comment]
Looks good!

Keywordless search (test) gives a single result titled `No results returned` with the magnet's info-hash being `btih:0000000000000000000000000000000000000000`.
[keywordless log.txt](https://github.com/Jackett/Jackett/files/5259336/keywordless.log.txt)

Also, the links for results direct to http://127.0.0.1:9117/UI/Dashboard (or whatever you use).

**Comment by garfield69** [comment]
do you know if the api has a recent list that can provide results for keywordless searches?
also, if the api provides a link to the torrent details page, then assigning it to release.Comments will be helpful.

**Comment by ilike2burnthing** [comment]
Recent list - https://apibay.org/precompiled/data_top100_recent.json

File: src/Jackett.Common/Indexers/ApiBay.cs
Comment (masterwok) [reply src/Jackett.Common/Indexers/ApiBay.cs:33]: Nice catch, thanks!

File: src/Jackett.Common/Indexers/ApiBay.cs
Comment (garfield69) [inline src/Jackett.Common/Indexers/ApiBay.cs:164]: I believe routine ParseDateTimeFromUnixEpochTimeStamp is a duplicate of an already existing tool, as shown in this example
https://github.com/Jackett/Jackett/blob/621a47338dcd98ff0b2471b7df99f7995e10a4dd/src/Jackett.Common/Indexers/HDBitsApi.cs#L119
please note I'm not a c# dev so you'll need to verify for yourself

Comment (masterwok) [reply src/Jackett.Common/Indexers/ApiBay.cs]: This appears to produce the same results in my testing. I've removed the duplicate method in favor of this one. Thanks!

**Comment by masterwok** [comment]
Thank you both for your feedback and help. I've made the following changes:

- Resolved `No results returned` issue (this was caused by the API returning a single item to represent no results)
- Updated `release.Comments` to point to `https://pirateproxy.cloud/description.php?id={id}` rather than localhost
- Keywordless search terms now return recent torrents from [https://apibay.org/precompiled/data_top100_recent.json](https://apibay.org/precompiled/data_top100_recent.json)
- Removed duplicate DateTime parse method from `ParseUtil` in favor of existing method

**Comment by garfield69** [comment]
just pending code compliance review from ngosang

**Review by garfield69 — APPROVED** [review]

**Comment by ilike2burnthing** [comment]
Works and looks well. Thanks for all the work.

I think apibay.org and pirateproxy.cloud need to be switched in terms of which is public and which is private, as it's pirateproxy.cloud that we want users to be able to change, not the link for the api. 

You could also add the other links from https://proxybay.gdn/ (pretty sure all but knaben.xyz will work) as AlternativeSiteLinks, e.g. - https://github.com/Jackett/Jackett/blob/master/src/Jackett.Common/Indexers/TorrentDay.cs#L24-L34

File: src/Jackett.Common/Indexers/ApiBay.cs
Comment (ngosang) [inline src/Jackett.Common/Indexers/ApiBay.cs:140]: remove braces when possible

File: src/Jackett.Common/Indexers/ApiBay.cs
Comment (ngosang) [inline src/Jackett.Common/Indexers/ApiBay.cs:163]: remove braces when possible

File: src/Jackett.Common/Indexers/ApiBay.cs
Comment (ngosang) [inline src/Jackett.Common/Indexers/ApiBay.cs:19]: remove this comment

File: src/Jackett.Common/Indexers/ApiBay.cs
Comment (ngosang) [inline src/Jackett.Common/Indexers/ApiBay.cs:21]: Add `ExcludeFromCodeCoverage` => https://github.com/Jackett/Jackett/blob/master/src/Jackett.Common/Indexers/YTS.cs#L19

File: src/Jackett.Common/Indexers/ApiBay.cs
Comment (ngosang) [inline src/Jackett.Common/Indexers/ApiBay.cs:198]: Add Files = num_files

Comment (ngosang) [inline src/Jackett.Common/Indexers/ApiBay.cs:198]: Did you test some results with IMBD id?
You can search `harry` to try it.

File: src/Jackett.Common/Indexers/ApiBay.cs
Comment (ngosang) [inline src/Jackett.Common/Indexers/ApiBay.cs:154]: Use `NameValueCollection` object to build the URL for you. It will use URL encode for the parameters and the code is more readable. => https://github.com/Jackett/Jackett/blob/master/src/Jackett.Common/Indexers/YTS.cs#L101

**Review by ngosang — CHANGES_REQUESTED** [review]
Nice job!!

**Comment by ngosang** [comment]
@garfield69 @ilike2burnthing This can be merged after the requested changes are done.

Since the official domain "thepiratebay.org" is using the JSON API, I think the HTML proxies are using an outdated version of the site. We should replace the old indexer with this one (same id so users don't have to re-configure the tracker).
The list of alternative sites should be configurable but there is an issue: the description domain and the API domain may be different.
IMHO we should use alternative sites for the description domain (is the domain the user can see in his browser) thepiratebay.org, pirateproxy.cloud ...
And add a new text field for the API domain: apibay.org We can provide alternative API domains in an info message like => https://pirateproxy.cloud/newapi/q.php?q=harry&cat=
All of this should be done in other PR to not block this.

**Comment by garfield69** [comment]
>And add a new text field for the API domain: apibay.org 

I disagree with this. 
Adding a text field in the config for a domain leads to issue with not being able to auto-update it when it requires changing. We are then dependent on users updating the config manually.
For an example of how messy this can get just look at the ongoing issues  with keeping YGGTorrent up to date.

The api domain can remain internal, and if it changes then we update the indexer C# and that will get the job done.

**Comment by garfield69** [comment]
I've switched to a "replace TPB yml with c#" and added supporting changes
only thing left is the NameValueCollection changes which I'll leave to you.
I've tested all my changes in the 0.16.1334 PR  build

**Comment by ilike2burnthing** [comment]
Do we need to remove thepiratebay.yml as well?

I figured knaben.xyz wouldn't work, it uses a different URL structure, so we can bump that one off. I also can't get the onion links to work, but I don't know if they're just being temperamental at the moment.

**Comment by garfield69** [comment]
>Do we need to remove thepiratebay.yml as well?

its part of the PR now

>I figured knaben.xyz wouldn't work, it uses a different URL structure, so we can bump that one off. I also can't get the onion links to work, but I don't know if they're just being temperamental at the moment.

knaben.xyz redirected to *.ru 
I tested all the alternatelinks with the new indexer and the TEST accepted them all.

**Comment by garfield69** [comment]
>I figured knaben.xyz wouldn't work, it uses a different URL structure, so we can bump that one off. I also can't get the onion links to work, but I don't know if they're just being temperamental at the moment.

Oh, I see now. Oversight on my part, was testing only the API duh!
yes, I will test that the description link works for all the alternate links and drop the ones that dont work

**Comment by ilike2burnthing** [comment]
Oh sorry, I meant delete src/Jackett.Common/Definitions/thepiratebay.yml from our repo, as we only have Program.cs deleting it for the user.

Knaben is a very weird proxy; its homepage is .ru, the links on the homepage are .xyz, and when you run a search it is .net. Then to add to the mess, our description link would be https://knaben.ru/description.php?id=36570088 but it 404s out (doesn't matter what domain you use), and the actual link for that result is https://knaben.net/s/keepgoing.php?url=/torrent/36570088/The_Secret_Life_of_a_Celebrity_Surrogate_2020_Lifetime_720p_HDTV

**Comment by garfield69** [comment]
>I meant delete src/Jackett.Common/Definitions/thepiratebay.yml 

OH Duh! Thanks I forgot. Now done in the PR.

**Comment by ilike2burnthing** [comment]
Well I'm sure you'll be very surprised to hear that 0.16.1336 also works fine (on Windows anyway) 😋

File: src/Jackett.Common/Indexers/ApiBay.cs
Comment (masterwok) [reply src/Jackett.Common/Indexers/ApiBay.cs]: Thank you, just pushed up this change.

**Comment by ilike2burnthing** [comment]
Working on Windows and enhanced log shows no errors, nice job all, particularly masterwok! 

We'll let @ngosang sign off on the changes and then we're probably good to go.

If we wanted 100 results for the test, adding the second page would get us that (50 per page) - https://apibay.org/precompiled/data_top100_recent_1.json

**Comment by masterwok** [comment]
Thanks everyone! :)

**Pull request merged**
